shaders/src/common_instancing.glsl

```diff
@@ -17,45 +21,35 @@ void initObjectUniforms() {
     //   on fields of structs, unless they're in a UBO (which we just copied out of).
 
 #if defined(FILAMENT_HAS_FEATURE_INSTANCING)
-#if defined(MATERIAL_HAS_INSTANCES)
-    if ((objectUniforms.data[0].flagsChannels & FILAMENT_OBJECT_INSTANCE_BUFFER_BIT) == 0) {
-        // the material manages instancing, all instances share the same uniform block.
-        object_uniforms_worldFromModelMatrix        = objectUniforms.data[0].worldFromModelMatrix;
-        object_uniforms_worldFromModelNormalMatrix  = objectUniforms.data[0].worldFromModelNormalMatrix;
-        object_uniforms_morphTargetCount            = objectUniforms.data[0].morphTargetCount;
-        object_uniforms_flagsChannels               = objectUniforms.data[0].flagsChannels;
-        object_uniforms_objectId                    = objectUniforms.data[0].objectId;
-        object_uniforms_userData                    = objectUniforms.data[0].userData;
-    } else
-#endif
-    {
-        // the object has an instance buffer
-        object_uniforms_worldFromModelMatrix        = objectUniforms.data[instance_index].worldFromModelMatrix;
-        object_uniforms_worldFromModelNormalMatrix  = objectUniforms.data[instance_index].worldFromModelNormalMatrix;
-        object_uniforms_morphTargetCount            = objectUniforms.data[instance_index].morphTargetCount;
-        object_uniforms_flagsChannels               = objectUniforms.data[instance_index].flagsChannels;
-        object_uniforms_objectId                    = objectUniforms.data[instance_index].objectId;
-        object_uniforms_userData                    = objectUniforms.data[instance_index].userData;
+    highp int i;
+#   if defined(MATERIAL_HAS_INSTANCES)
+    // instancing handled by the material
+    if ((objectUniforms.data[0].flagsChannels & FILAMENT_OBJECT_INSTANCE_BUFFER_BIT) != 0) {
+        // hybrid instancing, we have a instance buffer per object
+        i = instance_index;
+    } else {
+        // fully manual instancing
+        i = 0;
     }
+#   else
+    // automatic instancing
```

Comment:
This could be hybrid-instancing too, if the material doesn't specify `instanced: true`.